mobile/lib/features/trade/position_change_notifier.dart

```diff
@@ -37,12 +38,29 @@ class PositionChangeNotifier extends ChangeNotifier implements Subscriber {
     if (event is bridge.Event_PositionUpdateNotification) {
       Position position = Position.fromApi(event.field0);
 
-      position.unrealizedPnl = Amount(_positionService.calculatePnl(position, _bid, _ask));
-
+      if (_price != null) {
+        position.unrealizedPnl =
+            Amount(_positionService.calculatePnl(position, _price!));
+      } else {
+        position.unrealizedPnl = null;
+      }
       positions[position.contractSymbol] = position;
     } else if (event is bridge.Event_PositionClosedNotification) {
-      ContractSymbol contractSymbol = ContractSymbol.fromApi(event.field0.contractSymbol);
+      ContractSymbol contractSymbol =
+          ContractSymbol.fromApi(event.field0.contractSymbol);
       positions.remove(contractSymbol);
+    } else if (event is bridge.Event_PriceUpdateNotification) {
+      _price = Price.fromApi(event.field0);
+      for (ContractSymbol symbol in positions.keys) {
+        if (_price != null) {
```

Comment:
FWIW, I find it harder to read, so I guess it's subjective - for me it's more natural to go "if price is not empty, do ....). otherwise it jumps in a weird way for me. 

also: it's not really an invariant, it's just a conditional?

Reply:
oh, maybe we're using slightly different definitions of invariant.
for me, this word means "never changing" - so you can for example assert somethings (e.g. this will always have a value).
in this case, an invariant would be the following: `assert_eq(price != null)` 

(it's a very common pattern in C++ world to state the invariants, like "this pointer can't be null" , because there's just so many footguns....)

Reply:
your definition is probably more correct 👍